File: 496.下一个更大元素I.go

// monotonic stack & hash map
func nextGreaterElement(subset []int, nums []int) []int {
	stk := []int {}
	m := map[int]int {}
	// construct a monotinic stack
	for i := len(nums) - 1; i >= 0; i-- {
		// guarantee to choose the greater element to the right of the current
		for len(stk) > 0 && stk[len(stk) - 1] < nums[i] {
			stk = stk[: len(stk) - 1]
		}
		if len(stk) > 0 {
			m[nums[i]] = stk[len(stk) - 1]
		} else {
			m[nums[i]] = -1
		}
		stk = append(stk, nums[i])
	}
	ans := make([]int, len(subset))
	for i, num := range subset {
		ans[i] = m[num]
	}
	return ans
}

File: 496.下一个更大元素 I.cpp
#include <vector>
#include <stack>
#include <unordered_map>

using std::vector;
using std::stack;
using std::unordered_map;


class Solution {
public:
    // monotinic stack & hash map
    vector<int> nextGreaterElement(vector<int>& subset, vector<int>& nums) {
        // construct a monotonic stack with push & pop
        for (const int& i : nums) {
            while (stk.size() && stk.top() < i) {
                m[stk.top()] = i;
                stk.pop();
            }
            stk.push(i);
        }
        int len = subset.size();
        vector<int> ans(len, -1);
        for (int i = 0; i < len; ++i) {
            auto itr = m.find(subset[i]);
            ans[i] = itr == m.end() ? -1 : itr->second;
        }
        return ans;
    }

private:
    stack<int> stk;
    unordered_map<int, int> m;
};